### PR TITLE
Create new base images to use for pro product images

### PR DESCRIPTION
**Pull request by ianpittwood**
The [`r-docker/r-base`](https://github.com/rstudio/r-docker) images are very minimal and only install a singular version of R and a few core packages. This change adds new base images with and without the RStudio Pro Drivers included.These base images are very comprehensive compared to `r-base` and allow our actual professional product Dockerfiles to be slimmed down to only a few `RUN` statements used to install and configure the product itself.

This change adds or modifies the following:
- Six new base image options:
  - `base/` provides the base set of system dependencies, two versions of R, two versions of Python, tini, TInyTeX, and Quarto.
  - `base/pro` provides the same content as `base/` but with the addition of the RStudio Pro Drivers.
  - Support for three OSes: Ubuntu 18.04 (re-stylized `ubuntu1804`), Ubuntu 22.04 (re-stylized `ubuntu2204`), and CentOS7.
    - I chose to change the stylization of the Ubuntu OS names, which previously used the release codename, to use the version number instead. Best practices from other organizations (such as Snyk) push towards using the version number instead of the codename since it's easier to comprehend. In the new style it's also much more obvious which OS is being used at a glance if the viewer is not familiar with Ubuntu's release names.
- Professional product Dockerfile definitions using the new base images (i.e. `connect/Dockerfile.ubuntu1804.draft`).
- A Justfile for the base images directory.
- Goss tests for both the `base` and `base/pro` images.
- A new job in `build-latest.yaml` to build the base images prior to building the professional products. This will allow us to begin building and testing the base images from the Github Workflow before we fully implement them into the product images.
- Linting for all base image Dockerfiles.
- New targets in `ci.Justfile` for building and testing the base images.

Remaining work once this change is merged:
- Fully testing and implementing the base images into the pro product Dockerfiles.
- Removing the `content` images since these new base images will likely take their place.

This change **will not** change the end user experience as we will not yet push any product images that use the new base images.

### CODE REVIEW

**Comment by ianpittwood** [comment]
I'm still working on the CI, but I'm putting this up for review now to get feedback on the images themselves.

File: base/Dockerfile.ubuntu1804 [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        sudo \
+        tk \
+        tk-dev \
+        tk-table \
+        tzdata \
+        unixodbc-dev \
+        wget \
+        zlib1g-dev
+
+### Install tini ###
```
[the remark this answers is not in the document]
Comment (colearendt) [reply base/Dockerfile.ubuntu1804:79]: Great thought! We did something similar for `wait-for-it`: https://github.com/rstudio/rstudio-docker-products/pull/344

**Review by colearendt — CHANGES_REQUESTED** [review]
Haven't dug in super deep on this one yet. Although I definitely do appreciate the direction! 

A couple of immediate questions I have:
- Were the `content-base` and `content-pro` images considered for usage here? In spirit, definition, and practice these images seem to be a very close redevelopment of those images with some slight ordering / sysdep differences. I mentioned this [here](https://github.com/rstudio/rstudio-docker-products/issues/391#issuecomment-1256582714)
- One noteworthy difference is one version of R/Python versus two. That seems like something that should not necessarily warrant two completely different dockerfiles / approaches though 🤷 
    - In contrast, a benefit of consistency here is that it would make it much easier for someone to build a different RSW/RSC/RSPM from a different base image by just tweaking the args to the product build
- Is there a reason we changed from the codenames to version names? i.e. `bionic` to `ubuntu1804`?
- I also think it's worth noting the tagging convention in use for discussion
```
FROM rstudio/base-pro:ubuntu2204-r${R_VERSION}_${R_VERSION_ALT}-py${PYTHON_VERSION}_${PYTHON_VERSION_ALT}
```
- Is the plan for this PR to leave the drafts AS-IS, or to merge them with the other `Dockerfile`s? (My preference would definitely be to follow-up with overwriting later as a separate PR). 
    - One piece of this: the follow-up may be a good time to consider the pattern for including drivers in service images: whether we want to include them all the time, etc. They are awfully large, but in the spirit of our "kitchen sink" w/ lots of R, Python, etc., perhaps they are a good fit.
    - It will also be important to note any sysdep changes that _ultimately_ result inside of the product images. I presume care has been taken there, but it will be a tricky thing for someone to review 😄 
- It's also worth noting the complexity this introduces for customers who are copying our code to build their own images (i.e. they cannot just "pull" images themselves), or to customers who are already doing that and will have to go through a refactor to change their build process
    - perhaps documentation for this change in NEWS / etc. can wait til the follow-up as well. Just wanted to note here
- Should we be introducing `rstudio.` labels at this point? I suspect we will have many many changes to make with the change to `posit`, but labels seem like one that this may be a good time to start off on the other foot 😛 
    - In this vein, is this something we should land as a part of the rebrand, since it will be a pretty sizable change?
- One background point to highlight - we have a bit of an "open question" right now with respect to Quarto and how it will be versioned alongside R and Python. Right now you have a latent "one version of quarto without a tag referencing it" (i.e. version of quarto can float and is not an explicit dependency). 
    - https://github.com/rstudio/rstudio-docker-products/blob/main/content/base/bionic/maybe_install_quarto.sh (we only install it into some base images at present)
    - In one spirit, there is the "let quarto float" school of thought and maybe some convention for bumping the version of quarto when we add new R / Pythons. In the other vein, we could consider quarto a "top level" version and add it to the tag... which is already pretty long 😅

File: base/Dockerfile.ubuntu1804 [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+ADD https://github.com/krallin/tini/releases/download/v${TINI_VERSION}/tini-amd64 /tini
+ADD https://github.com/krallin/tini/releases/download/v${TINI_VERSION}/tini-amd64.asc /tini.asc
+RUN gpg --batch --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 595E85A6B1B4779EA4DAAEC70B588DFF0527A9B7 \
+    && gpg --batch --verify /tini.asc /tini \
+    && chmod +x /tini \
+    && ln -s /tini /usr/local/bin/tini
+
+### Install TinyTeX ###
+RUN curl -sL "https://yihui.org/tinytex/install-bin-unix.sh" | sh && \
```
[the remark this answers is not in the document]
Comment (colearendt) [reply base/Dockerfile.ubuntu1804]: On this one - does anyone know if TinyTex is included in our aims for `.deb`/`.rpm` repository?

**Comment by ianpittwood** [comment]
I used a separate text editor to try and make some responses for you here. Bear with me!

> Haven't dug in super deep on this one yet. Although I definitely do appreciate the direction!
> 
> A couple of immediate questions/comments I have:
> 
>     * Were the `content-base` and `content-pro` images considered for usage here? In spirit, definition, and practice these images seem to be a very close redevelopment of those images with some slight ordering / sysdep differences. I mentioned this [here](https://github.com/rstudio/rstudio-docker-products/issues/391#issuecomment-1256582714)
>       
>       * Sorry I missed your comment that these should replace the content images. Worth noting that the content images are used downstream / [referenced explicitly](https://github.com/rstudio/helm/blob/main/charts/rstudio-connect/default-runtime.yaml) by our Connect helm chart, customer infrastructure, etc. Is setting these images up as `base` a temporary naming convention? Or is the proposal to deprecate `content-base` as an image tag, and just go with `base` going forwards?

This is great feedback! I honestly wasn't aware that content base/pro were being used downstream for things. I do think that these changes should still *replace* the content images and here's my reasoning behind it:
* The content images are not comprehensive (i.e. missing alt versions of R and Python, no tini, etc.)
* We have a pattern to our other images. The content images do not match this pattern (`content/base/bionic/Dockerfile` vs `connect/Dockerfile.bionic`).

We can refactor the existing content directory and files, but I thought it was easier to just start from scratch.

>       * I seem to recall some discussion for why `content-base` was chosen over `base` when we first built the `content` images. Maybe someone else will recall better than I can. Perhaps because we considered the "content" (no pun intended) of the images to be such that "content" could run. I.e. it is not a base image that you would use for installing R, etc. It is a base image that you would use for building some type of content (i.e. flask app, plumber API, Connect image, etc.). i.e. in contrast to [these base images](https://github.com/rstudio/r-docker/tree/main/base)

That explanation for the name makes more sense to me now, but admittedly I don't love it. I'd prefer using `base` as the name since `content` is less obvious.

>     * One noteworthy difference is one version of R/Python versus two. That seems like something that should not necessarily warrant two completely different dockerfiles / approaches though shrug

Including two versions was an attempt to satisfy the requirements that the product images set out as all of them use a primary and alt version. I'd lean towards including more instead of less in a base image. I agree that it doesn't warrant two Dockerfiles.

>       * In contrast, a benefit of consistency here is that it would make it much easier for someone to build a different RSW/RSC/RSPM from a different base image by just tweaking the args to the product build
>       * What are y'all's thoughts on the `content-base` image having one R / Python version, but these images having two?
> 
>     * Is there a reason we changed from the codenames to version names? i.e. `bionic` to `ubuntu1804`?

I think using Ubuntu codenames is unhelpful to a lot of people less familiar with Linux. There are a lot of us that know the last 6 years of codenames off the top of our heads, but I know there's people that also need to google bionic to figure out which version that is. I think it's easier to use the OS's name and version number since that explicitly tells people that we're using Ubuntu 18.04 whereas Bionic only implies that. Tools like Snyk also push towards using explicit version numbers over code names so this change also satisfies those warnings.

>     * I also think it's worth noting the tagging convention in use for discussion
> 
> 
> ```
> FROM rstudio/base-pro:ubuntu2204-r${R_VERSION}_${R_VERSION_ALT}-py${PYTHON_VERSION}_${PYTHON_VERSION_ALT}
> ```
> 
>     * Recording some notes for the follow-up issues while they are front of mind.
>       
>       * One piece of this: the follow-up may be a good time to consider the pattern for including drivers in service images: whether we want to include them all the time, etc. They are awfully large, but in the spirit of our "kitchen sink" w/ lots of R, Python, etc., perhaps they are a good fit.

So here's my question: does anyone use a pro driver compatible product without the pro drivers? Since I'm still pretty new I'm not sure what kind of usage we see on that. Maybe we could control using the base or base-pro images for our pro product images with an `ARG`?

>       * It will also be important to note any sysdep changes that _ultimately_ result inside of the product images. I presume care has been taken there, but it will be a tricky thing for someone to review smile

This is certainly something I struggled with when building the base images. Do we have any SSOT on product dependencies that we could leverage here?

>     * It's also worth noting the complexity this introduces for customers who are copying our code to build their own images (i.e. they cannot just "pull" images themselves), or to customers who are already doing that and will have to go through a refactor to change their build process

Yeah it's certainly a trade off we have to weigh. A more comprehensive base image is much easier for us to maintain, but may introduce some new difficulties if folks are making their own images. Really though you could just copy both the base image and the product image into the same file and have a decently easy time modifying them for your own purposes.

>       * perhaps documentation for this change in NEWS / etc. can wait til the follow-up as well. Just wanted to note here
>       * In this vein, is this something we should land as a part of the rebrand, since it will be a pretty sizable change? To be clear, I mean "switching to use the base images" and not necessarily this PR

Sounds like a good plan to me.

>     * Should we be introducing `rstudio.` labels at this point? I suspect we will have many many changes to make with the change to `posit`, but labels seem like one that this may be a good time to start off on the other foot stuck_out_tongue

Fair point! We could maybe include both for now and remove the `rstudio.*` ones later? Or we could just use `posit` from the start since by the time this gets used we should be well into the rebrand.

>     * One background point to highlight - we have a bit of an "open question" right now with respect to Quarto and how it will be versioned alongside R and Python. Right now you have a latent "one version of quarto without a tag referencing it" (i.e. version of quarto can float and is not an explicit dependency).
>       
>       * https://github.com/rstudio/rstudio-docker-products/blob/main/content/base/bionic/maybe_install_quarto.sh (we only install it into some base images at present)
>       * #380
>       * In one school of thought, there is the "let quarto float" approach and maybe some convention for bumping the version of quarto when we add new R / Pythons. In another, we could consider quarto a "top level" version and add it to the tag... which is already pretty long sweat_smile
>       * For background, Connect treats Quarto more like R or Python. Further, the speed of development is partially a concern. i.e. what do we do when a new version of Quarto comes out that we want our containers to support out of the box, but no new R or Python has shipped. Do we introduce a breaking change? Do we bump quarto into the existing image tags and hope for the best but destabilize the image tags? etc.

Hmm that's much less straightforward than I had assumed. I'll give this some additional thought and circle back.

**Comment by kfeinauer** [comment]
> This is certainly something I struggled with when building the base images. Do we have any SSOT on product dependencies that we could leverage here?

Required product dependencies are typically spelled out within the DEB/RPM being installed for the product. For secondary/optional dependencies, I'm not sure how to get that information without digging into the development/Docker containers for product builds.

**Comment by colearendt** [comment]
Thanks for your responses! One discussion point I think we should land on:

- How many variations of this "base" image should there be? I.e. the `content` image currently has a huge matrix, because the images are meant to be a bit more surgical: https://github.com/rstudio/rstudio-docker-products/blob/main/content/matrix.json

And one other random point - the `content` images were actually consistent with everything until Daniel refactored things this summer. He did not tackle the content images, so they are still in the old world. I do think that `base` is too generic for an org-wide repository like `rstudio` or `posit`. Depending on the question above, it may be that we just want a `product-base` and `product-pro` image? `tini` is small, but it is an example of something that is needed for a "product" image, but _not_ for a "content" image AFAIK.... or maybe we should start using `tini` or something like it in content image execution? 🤷

**Comment by ianpittwood** [comment]
@colearendt I'll go through and update the `content` images quickly in a separate PR then so we can have those up to date.

Yeah I'm onboard with making these `product-base` since it's really serving that specific purpose. I'll rename stuff to match that.

File: base/Dockerfile.ubuntu1804 (hunk 2 above)
[the remark this answers is not in the document]
Comment (ianpittwood) [reply base/Dockerfile.ubuntu1804]: > Same question as above -- we should consider hosting these ourselves from a trusted location

I would probably categorize the current location as trusted. Yihui is an RStudio employee and the repository for TinyTeX is under our organization.

**Review by colearendt — APPROVED** [review]
I'm inclined to go ahead and merge this and fix forward any issues 😅 Kinda hard to see the pain points until we start wrestling with it. I'm thinking we will target `r-session-complete` or `workbench` as our next target for using the base images.

**Pull request merged**